src/components/text/text.styles.ts

```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { css } from '@emotion/react';
+import { UseEuiTheme } from '../../services';
+import {
+  logicalCSS,
+  euiFontSize,
+  _FontScaleOptions,
+} from '../../global_styling';
+
+import { euiLinkCSS } from '../link/link.styles';
+import { euiTitle } from '../title/title.styles';
+
+export const euiTextConstrainedMaxWidth = '36em';
+
+/**
+ * Mixins
+ */
+export const euiText = (
+  euiTheme: UseEuiTheme['euiTheme'],
+  inheritColor = false
+) => {
+  return {
+    color: inheritColor ? 'inherit' : euiTheme.colors.text,
+    fontWeight: euiTheme.font.weight.regular,
+  };
+};
+
+// Internal utility for EuiText scales/sizes
+const euiScaleText = (
+  euiTheme: UseEuiTheme['euiTheme'],
+  options: _FontScaleOptions
+) => {
+  const { fontSize, lineHeight } = euiFontSize('m', euiTheme, options);
+  const { measurement } = options;
+  const lineHeightSize = measurement === 'em' ? `${lineHeight}em` : lineHeight;
+
+  const headings = {
+    h1: euiTitle('l', euiTheme, options),
+    h2: euiTitle('m', euiTheme, options),
+    h3: euiTitle('s', euiTheme, options),
+    h4: euiTitle('xs', euiTheme, options),
+    h5: euiTitle('xxs', euiTheme, options),
+    h6: euiTitle('xxxs', euiTheme, options),
+  };
+  const headingMarginTop = (
+    headingFontSize: ReturnType<typeof euiTitle>['fontSize']
+  ) => {
+    let marginTop = parseFloat(String(fontSize)) * 2;
+    if (measurement === 'em') {
+      marginTop = marginTop / parseFloat(String(headingFontSize));
+    }
+    return `${marginTop}${measurement}`;
+  };
+  const headingMarginBottom = (
+    headingFontSize: ReturnType<typeof euiTitle>['fontSize']
+  ) => {
+    let marginBottom = parseFloat(String(fontSize)) / 2;
+    if (measurement === 'em') {
+      marginBottom = marginBottom / parseFloat(String(headingFontSize));
+    }
+    return `${marginBottom}${measurement}`;
+  };
+
+  return `
+    font-size: ${fontSize};
+    line-height: ${lineHeight};
+
+    h1 {
+      font-size: ${headings.h1.fontSize};
+      line-height: ${headings.h1.lineHeight};
+    }
+    h1:not(:last-child) {
+      ${logicalCSS('margin-bottom', headingMarginBottom(headings.h1.fontSize))}
+    }
+
+    h2 {
+      font-size: ${headings.h2.fontSize};
+      line-height: ${headings.h2.lineHeight};
+    }
+    h2:not(:last-child) {
+      ${logicalCSS('margin-bottom', headingMarginBottom(headings.h2.fontSize))}
+    }
```

Comment:
Optional nit: We can still nicely nest these selectors to reduce duplication.
```suggestion
    h2 {
      font-size: ${headings.h2.fontSize};
      line-height: ${headings.h2.lineHeight};
      &:not(:last-child) {
        ${logicalCSS('margin-bottom', headingMarginBottom(headings.h2.fontSize))}
      }
    }
```

Reply:
I kept it un-nested to keep it visually on the same indentation level as the `* + h2` below it. Unfortunately that selector can't be nested (the `+ h2` gets all messed up if swapped for `+ &`) and it felt like the two effects conceptually belonged together.

Reply:
Oh - or alternatively, maybe we could use `:not` instead of `+` to keep everything nested?

```
h2 {

  &:not(:first-child) {
    ${logicalCSS('margin-top', headingMarginTop(headings.h2.fontSize))}
  }

  &:not(:last-child) {
    ${logicalCSS('margin-bottom', headingMarginBottom(headings.h2.fontSize))}
  }
```

Reply:
Yeah I like sticking with `first` vs `last` better. It's more obvious and I think most likely not to apply erroneously.

Reply:
https://github.com/elastic/eui/pull/5895/commits/49739f7feeffa5d10592d02937463ee4ec49495b